roachtest: follower-reads/mixed-version/single-region failed

roachtest.follower-reads/mixed-version/single-region [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/5582417?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/5582417?buildTab=artifacts#/follower-reads/mixed-version/single-region) on master @ [7f8fef4c3f24ec2b6f9a3db1eb38756da68b3d5a](https://github.com/cockroachdb/cockroach/commits/7f8fef4c3f24ec2b6f9a3db1eb38756da68b3d5a):


```
test artifacts and logs in: /artifacts/follower-reads/mixed-version/single-region/run_1
	follower_reads.go:295,follower_reads.go:901,follower_reads.go:102,test_runner.go:896: error verifying node values: dial tcp 34.148.137.234:26257: connect: connection refused
```
<p>Parameters: <code>ROACHTEST_cloud=gce</code>
, <code>ROACHTEST_cpu=2</code>
, <code>ROACHTEST_ssd=0</code>
</p>
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*follower-reads/mixed-version/single-region.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-17109

roachtest.follower-reads/mixed-version/single-region [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/5594177?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/5594177?buildTab=artifacts#/follower-reads/mixed-version/single-region) on master @ [ac11a55499120ffd33016feebc47b4ed43a21007](https://github.com/cockroachdb/cockroach/commits/ac11a55499120ffd33016feebc47b4ed43a21007):


```
test artifacts and logs in: /artifacts/follower-reads/mixed-version/single-region/run_1
	follower_reads.go:295,follower_reads.go:901,follower_reads.go:102,test_runner.go:896: error verifying node values: dial tcp 35.227.105.156:26257: connect: connection refused
```
<p>Parameters: <code>ROACHTEST_cloud=gce</code>
, <code>ROACHTEST_cpu=2</code>
, <code>ROACHTEST_ssd=0</code>
</p>
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*follower-reads/mixed-version/single-region.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

https://teamcityartifacts.crdb.io/presignedTokenAuth/a04e0433-e0f3-4d1a-8de3-b959241077df/repository/download/Cockroach_Nightlies_RoachtestNightlyGceBazel/5594177:id/follower-reads/mixed-version/single-region/run_1/artifacts.zip!/logs/1.unredacted/cockroach-stderr.log

I would hazard a guess that this is invalid span usage (vs a bug in tracing), and it seems to be coming from a trace span thati s set up and finished as part of the jobs infrastructure. cc @cockroachdb/jobs 

```
I220629 09:55:46.609531 1 util/log/flags.go:198  [-] 1  stderr capture started
panic: RecordingStructured has 12 recordings; expected 1 [recovered]
	panic: RecordingStructured has 12 recordings; expected 1 [recovered]
	panic: RecordingStructured has 12 recordings; expected 1

goroutine 831 [running]:
panic({0x46f6f80, 0xc003530960})
	GOROOT/src/runtime/panic.go:1147 +0x3a8 fp=0xc001d5ff88 sp=0xc001d5fec8 pc=0x48a828
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:58 +0x3bb fp=0xc001d60120 sp=0xc001d5ff88 pc=0x1ffd71b
runtime.deferCallSave(0xc001d60228, 0xc001d61dc8)
	GOROOT/src/runtime/panic.go:950 +0x82 fp=0xc001d60130 sp=0xc001d60120 pc=0x48a422
runtime.runOpenDeferFrame(0xc0029ac500, 0xc0029ac550)
	GOROOT/src/runtime/panic.go:889 +0x27b fp=0xc001d601b0 sp=0xc001d60130 pc=0x489dbb
panic({0x46f6f80, 0xc003530960})
	GOROOT/src/runtime/panic.go:1038 +0x215 fp=0xc001d60270 sp=0xc001d601b0 pc=0x48a695
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:58 +0x3bb fp=0xc001d60408 sp=0xc001d60270 pc=0x1ffd71b
runtime.deferCallSave(0xc001d60510, 0xc001d61cf0)
	GOROOT/src/runtime/panic.go:950 +0x82 fp=0xc001d60418 sp=0xc001d60408 pc=0x48a422
runtime.runOpenDeferFrame(0xc0029ac550, 0xc0029ac500)
	GOROOT/src/runtime/panic.go:889 +0x27b fp=0xc001d60498 sp=0xc001d60418 pc=0x489dbb
panic({0x46f6f80, 0xc003530960})
	GOROOT/src/runtime/panic.go:1038 +0x215 fp=0xc001d60558 sp=0xc001d60498 pc=0x48a695
github.com/cockroachdb/cockroach/pkg/util/tracing.(*crdbSpan).recordFinishedChildrenLocked(0xc001f43558, {0xc001ef6900, 0x69708b8c99, 0xc001d60648})
	github.com/cockroachdb/cockroach/pkg/util/tracing/crdbspan.go:550 +0x325 fp=0xc001d605d8 sp=0xc001d60558 pc=0xbf3c25
github.com/cockroachdb/cockroach/pkg/util/tracing.(*crdbSpan).childFinished(0xc001f43558, 0xc001262e58)
	github.com/cockroachdb/cockroach/pkg/util/tracing/crdbspan.go:1069 +0x251 fp=0xc001d60658 sp=0xc001d605d8 pc=0xbf6bd1
github.com/cockroachdb/cockroach/pkg/util/tracing.(*crdbSpan).finish(0xc001262e58)
	github.com/cockroachdb/cockroach/pkg/util/tracing/crdbspan.go:279 +0x153 fp=0xc001d606e0 sp=0xc001d60658 pc=0xbf2513
github.com/cockroachdb/cockroach/pkg/util/tracing.(*spanInner).Finish(0xc001262e08)
	github.com/cockroachdb/cockroach/pkg/util/tracing/span_inner.go:101 +0x45 fp=0xc001d60710 sp=0xc001d606e0 pc=0xbfa505
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).finishInternal(0xc001262e00)
	github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:263 +0xb7 fp=0xc001d60738 sp=0xc001d60710 pc=0xbf7937
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).Finish(0xc001f43500)
	github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:251 +0x19 fp=0xc001d60750 sp=0xc001d60738 pc=0xbf7859
github.com/cockroachdb/cockroach/pkg/jobs.(*Job).load·dwrap·12()
	github.com/cockroachdb/cockroach/pkg/jobs/jobs.go:803 +0x26 fp=0xc001d60768 sp=0xc001d60750 pc=0x21735a6
github.com/cockroachdb/cockroach/pkg/jobs.(*Job).load(0xc00297b860, {0x6704458, 0xc004bb2b10}, 0xc0048e2480)
	github.com/cockroachdb/cockroach/pkg/jobs/jobs.go:853 +0x271 fp=0xc001d60830 sp=0xc001d60768 pc=0x2172eb1
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).LoadJobWithTxn(0xc000ada280, {0x6704458, 0xc004bb2b10}, 0xac05ce538c38001, 0x3ff8000000000000)
	github.com/cockroachdb/cockroach/pkg/jobs/registry.go:647 +0x7f fp=0xc001d60870 sp=0xc001d60830 pc=0x217b63f
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).LoadJob(...)
	github.com/cockroachdb/cockroach/pkg/jobs/registry.go:620
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).waitForJobs(0xc000ada280, {0x6704458, 0xc004bb2b10}, {0x67a4998, 0xc000f02ba0}, {0xc00505c4f0, 0x1, 0x1}, 0xc003c62b40)
	github.com/cockroachdb/cockroach/pkg/jobs/wait.go:121 +0x5d3 fp=0xc001d60cc0 sp=0xc001d60870 pc=0x218be73
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).Run(0xc002395c00, {0x6704458, 0xc004bb2b10}, {0x67a4998, 0xc000f02ba0}, {0xc00505c4f0, 0x0, 0x1})
	github.com/cockroachdb/cockroach/pkg/jobs/wait.go:174 +0x14e fp=0xc001d60d50 sp=0xc001d60cc0 pc=0x218cbce
github.com/cockroachdb/cockroach/pkg/upgrade/upgrademanager.(*Manager).runMigration(0xc0027e1100, {0x6704458, 0xc004bb2b10}, {{0x4fabac1, 0xc004bb2b10}}, {{0x2, 0x0, 0x1, 0x0}})
	github.com/cockroachdb/cockroach/pkg/upgrade/upgrademanager/manager.go:304 +0x154 fp=0xc001d60dd0 sp=0xc001d60d50 pc=0x38166d4
github.com/cockroachdb/cockroach/pkg/upgrade/upgrademanager.(*Manager).Migrate(0xc0027e1100, {0x67043b0, 0xc0048494c0}, {{0x4fabac1, 0xc004bb02d0}}, {{0xa, 0x0, 0x10, 0x0}}, {{0x16, ...}}, ...)
	github.com/cockroachdb/cockroach/pkg/upgrade/upgrademanager/manager.go:137 +0x505 fp=0xc001d60f08 sp=0xc001d60dd0 pc=0x3815c85
github.com/cockroachdb/cockroach/pkg/upgrade/upgrademanager.(*Manager).Migrate-fm({0x67043b0, 0xc0048494c0}, {{0x4fabac1, 0x9f26c40}}, {{0x4bb2ae0, 0xc0, 0x0, 0x0}}, {{0x16, 0x1, ...}}, ...)
	github.com/cockroachdb/cockroach/pkg/upgrade/upgrademanager/manager.go:99 +0x8b fp=0xc001d60f68 sp=0xc001d60f08 pc=0x38c924b
github.com/cockroachdb/cockroach/pkg/sql.runMigrationsAndUpgradeVersion({0x67043b0, 0xc0048494c0}, 0xc0004e2480, {{0x4fabac1, 0x0}}, {0x67d04a8, 0xc004a57540}, {0x67d04a8, 0xc004975b30}, 0xc004bb2ae0)
	github.com/cockroachdb/cockroach/pkg/sql/set_cluster_setting.go:585 +0x156 fp=0xc001d60fe8 sp=0xc001d60f68 pc=0x3369dd6
github.com/cockroachdb/cockroach/pkg/sql.setVersionSetting({0x67043b0, 0xc0048494c0}, 0xc0004e2480, 0xc000104a50, {0x513000f, 0xc0007160c0}, 0x0, {{0x4fabac1, 0x0}}, 0xc000e0a000, ...)
	github.com/cockroachdb/cockroach/pkg/sql/set_cluster_setting.go:521 +0x505 fp=0xc001d61170 sp=0xc001d60fe8 pc=0x3369025
github.com/cockroachdb/cockroach/pkg/sql.writeNonDefaultSettingValue({0x67043b0, 0xc0048494c0}, 0xc0004e2480, {0x67a9e08, 0xc000104a50}, {0x513000f, 0x7}, 0xc00451fad0, {{0x4fabac1, 0x4}}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/set_cluster_setting.go:384 +0x1db fp=0xc001d612e0 sp=0xc001d61170 pc=0x336873b
github.com/cockroachdb/cockroach/pkg/sql.writeSettingInternal.func1({0x67043b0, 0xc0048494c0}, 0x6782f68)
	github.com/cockroachdb/cockroach/pkg/sql/set_cluster_setting.go:320 +0x169 fp=0xc001d613d0 sp=0xc001d612e0 pc=0x3368189
github.com/cockroachdb/cockroach/pkg/kv.runTxn.func1({0x67043b0, 0xc0048494c0}, 0x4c1af3)
	github.com/cockroachdb/cockroach/pkg/kv/db.go:966 +0x27 fp=0xc001d613f8 sp=0xc001d613d0 pc=0x12f1887
github.com/cockroachdb/cockroach/pkg/kv.(*Txn).exec(0xc004531c20, {0x67043b0, 0xc0048494c0}, 0xc001d61738)
	github.com/cockroachdb/cockroach/pkg/kv/txn.go:959 +0xae fp=0xc001d616e0 sp=0xc001d613f8 pc=0x12f9ece
github.com/cockroachdb/cockroach/pkg/kv.runTxn({0x67043b0, 0xc0048494c0}, 0xc000714bd0, 0xddbb10)
	github.com/cockroachdb/cockroach/pkg/kv/db.go:965 +0x5a fp=0xc001d61758 sp=0xc001d616e0 pc=0x12f179a
github.com/cockroachdb/cockroach/pkg/kv.(*DB).TxnWithAdmissionControl(0x4ecca9, {0x67043b0, 0xc0048494c0}, 0x9f808d8, 0x0, 0xc000ddbad0)
	github.com/cockroachdb/cockroach/pkg/kv/db.go:928 +0x89 fp=0xc001d617a0 sp=0xc001d61758 pc=0x12f1529
github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn(...)
	github.com/cockroachdb/cockroach/pkg/kv/db.go:907
github.com/cockroachdb/cockroach/pkg/sql.writeSettingInternal({0x67043b0, 0xc0048494c0}, 0x10, {0x67a9e08, 0xc000104a50}, {0x513000f, 0x45df95}, {{0x4fabac1, 0x49eb2e0}}, 0xc000e0a000, ...)
	github.com/cockroachdb/cockroach/pkg/sql/set_cluster_setting.go:305 +0x145 fp=0xc001d61860 sp=0xc001d617a0 pc=0x3367f85
github.com/cockroachdb/cockroach/pkg/sql.(*setClusterSettingNode).startExec(0xc0048491c0, {{0x67043b0, 0xc0048494c0}, 0xc004a1a380, 0xc004936d08})
	github.com/cockroachdb/cockroach/pkg/sql/set_cluster_setting.go:215 +0x3df fp=0xc001d61990 sp=0xc001d61860 pc=0x336765f
github.com/cockroachdb/cockroach/pkg/sql.startExec.func2({0xc004a29a80, 0x4625b0}, {0x67068a8, 0xc0048491c0})
	github.com/cockroachdb/cockroach/pkg/sql/plan.go:515 +0x147 fp=0xc001d61a40 sp=0xc001d61990 pc=0x3302887
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal.func1()
	github.com/cockroachdb/cockroach/pkg/sql/walk.go:112 +0x3e fp=0xc001d61a78 sp=0xc001d61a40 pc=0x33bd75e
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal(0xc004a29b90, {0x67068a8, 0xc0048491c0}, {0x4fee9c9, 0x13})
	github.com/cockroachdb/cockroach/pkg/sql/walk.go:297 +0x14c9 fp=0xc001d61b08 sp=0xc001d61a78 pc=0x33bd669
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visit(0xc004a29b90, {0x67068a8, 0xc0048491c0})
	github.com/cockroachdb/cockroach/pkg/sql/walk.go:79 +0xf7 fp=0xc001d61b58 sp=0xc001d61b08 pc=0x33bc097
github.com/cockroachdb/cockroach/pkg/sql.walkPlan(...)
	github.com/cockroachdb/cockroach/pkg/sql/walk.go:43
github.com/cockroachdb/cockroach/pkg/sql.startExec({{0x67043b0, 0xc0048494c0}, 0xc004a1a380, 0xc004936d08}, {0x67068a8, 0xc0048491c0})
	github.com/cockroachdb/cockroach/pkg/sql/plan.go:518 +0x125 fp=0xc001d61bd8 sp=0xc001d61b58 pc=0x33026e5
github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Start(0xc004a20000, {0x67043b0, 0xc0048494c0})
	github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go:147 +0x87 fp=0xc001d61c18 sp=0xc001d61bd8 pc=0x33057c7
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Columnarizer).Init(0xc004a10c00, {0x67043b0, 0xc0048494c0})
	github.com/cockroachdb/cockroach/pkg/sql/colexec/columnarizer.go:159 +0xd4 fp=0xc001d61c58 sp=0xc001d61c18 pc=0x2800c94
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).Init(0xc004a22380, {0x67043b0, 0xc0048494c0})
	github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:90 +0x37 fp=0xc001d61c88 sp=0xc001d61c58 pc=0x3101577
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedStatsCollectorImpl).Init(0xc004a22380, {0x67043b0, 0xc0048494c0})
	<autogenerated>:1 +0x33 fp=0xc001d61cb0 sp=0xc001d61c88 pc=0x310cfb3
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).Start.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:250 +0x34 fp=0xc001d61cd8 sp=0xc001d61cb0 pc=0x2803ff4
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError(0x0)
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:91 +0x62 fp=0xc001d61d18 sp=0xc001d61cd8 pc=0x1ffd302
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).Start(0xc00452b680, {0x67043b0, 0xc0048494c0})
	github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:249 +0x78 fp=0xc001d61d88 sp=0xc001d61d18 pc=0x2803f38
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*FlowCoordinator).Start.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:122 +0x34 fp=0xc001d61db0 sp=0xc001d61d88 pc=0x30fd734
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError(0x0)
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:91 +0x62 fp=0xc001d61df0 sp=0xc001d61db0 pc=0x1ffd302
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*FlowCoordinator).Start(0xc00175e360, {0x67043b0, 0xc0048494c0})
	github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:121 +0x6b fp=0xc001d61e50 sp=0xc001d61df0 pc=0x30fd6ab
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBaseNoHelper).Run(0xc00175e360, {0x67043b0, 0xc0048494c0})
	github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:728 +0x3a fp=0xc001d61e90 sp=0xc001d61e50 pc=0x21d18ba
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*FlowCoordinator).Run(0xc00175e360, {0x67043b0, 0xc0048494c0})
	<autogenerated>:1 +0x33 fp=0xc001d61eb8 sp=0xc001d61e90 pc=0x310ad13
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run(0xc00452b4a0, {0x67043b0, 0xc0048494c0}, 0xc00175e360)
	github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:475 +0x258 fp=0xc001d61f78 sp=0xc001d61eb8 pc=0x26b2d18
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlow).Run(0xc0048ddb30, {0x67043b0, 0xc0048494c0}, 0x1)
	github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:249 +0x205 fp=0xc001d62018 sp=0xc001d61f78 pc=0x3102b65
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc0025bea00, {0x6704458, 0xc0048ddad0}, 0xc00493bce0, 0xc004531b80, 0xc004940b80, 0xc002de0000, 0xc004936fc8, 0x0)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:607 +0xb04 fp=0xc001d62768 sp=0xc001d62018 pc=0x327c884
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun(0x6704458, {0x6704458, 0xc0048ddad0}, 0x1, 0xc00493bce0, 0xc0048dd860, {{0x67068a8, 0xc0048491c0}, 0x0}, 0xc002de0000)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1455 +0x1b2 fp=0xc001d62820 sp=0xc001d62768 pc=0x3280672
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0xc004936600, {0x6704458, 0xc0048ddad0}, 0xc004936d08, 0xc0048ddad0, {0x67ac818, 0xc0048dd860}, 0x45, 0xc004977c18)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1485 +0x614 fp=0xc001d62998 sp=0xc001d62820 pc=0x320a614
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc004936600, {0x6704458, 0xc0048dd9b0}, 0xc004936d08, {0x67ac818, 0xc0048dd860})
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1159 +0xb87 fp=0xc001d62d98 sp=0xc001d62998 pc=0x32081a7
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc004936600, {0x67043b0, 0xc004849140}, {{0x6734e78, 0xc00497bda0}, {0x512fffb, 0x45}, 0x0, 0x0}, 0x0, ...)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:690 +0x2091 fp=0xc001d63670 sp=0xc001d62d98 pc=0x3204071
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc004936600, {0x67043b0, 0xc004849140}, {{0x6734e78, 0xc00497bda0}, {0x512fffb, 0x45}, 0x0, 0x0}, 0x0, ...)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:145 +0x59e fp=0xc001d638e0 sp=0xc001d63670 pc=0x320113e
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1({{{0x6734e78, 0xc00497bda0}, {0x512fffb, 0x45}, 0x0, 0x0}, {0xc0a7264a74b61db1, 0x1c5707f19, 0x0}, {0xc0a7264a74b61db1, ...}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1892 +0x2f6 fp=0xc001d63998 sp=0xc001d638e0 pc=0x31f6ab6
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0xc004936600)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1896 +0xb48 fp=0xc001d63e60 sp=0xc001d63998 pc=0x31f5048
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc004936600, {0x6704458, 0xc0048dd470}, 0x0, {0x0, 0x0, 0x0, 0x0, 0x0}, 0x0)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1818 +0x26c fp=0xc001d63f30 sp=0xc001d63e60 pc=0x31f412c
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).initConnEx.func1()
	github.com/cockroachdb/cockroach/pkg/sql/internal.go:206 +0xa5 fp=0xc001d63fe0 sp=0xc001d63f30 pc=0x32cbbc5
runtime.goexit()
	GOROOT/src/runtime/asm_amd64.s:1581 +0x1 fp=0xc001d63fe8 sp=0xc001d63fe0 pc=0x4bf321
created by github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).initConnEx
	github.com/cockroachdb/cockroach/pkg/sql/internal.go:205 +0x5f1
```

--- a/pkg/util/tracing/crdbspan.go
+++ b/pkg/util/tracing/crdbspan.go
@@ -546,12 +546,11 @@ func (s *crdbSpan) recordFinishedChildrenLocked(childRecording tracingpb.Recordi
 		// records by falling through.
 		fallthrough
 	case tracingpb.RecordingStructured:
-		if len(childRecording) != 1 {
-			panic(fmt.Sprintf("RecordingStructured has %d recordings; expected 1", len(childRecording)))
-		}
-
-		for i := range rootChild.StructuredRecords {
-			s.recordInternalLocked(&rootChild.StructuredRecords[i], &s.mu.recording.structured)
+		for ci := range childRecording {
+			child := &childRecording[ci]
+			for i := range child.StructuredRecords {
+				s.recordInternalLocked(&child.StructuredRecords[i], &s.mu.recording.structured)
+			}
 		}
 	case tracingpb.RecordingOff:
 		break

--- a/pkg/util/tracing/span_test.go
+++ b/pkg/util/tracing/span_test.go
@@ -410,6 +410,63 @@ func TestRecordingMaxSpans(t *testing.T) {
 	require.Len(t, root.StructuredRecords, extraChildren)
 }
 
+// TestRecordingDowngradesToStructuredIfTooBig finishes a span that has reached
+// the maximum number of recorded spans and asserts that its structured
+// recordings are correctly added to the parent.
+func TestRecordingDowngradesToStructuredIfTooBig(t *testing.T) {
+	now := timeutil.Now()
+	clock := timeutil.NewManualTime(now)
+	tr := NewTracerWithOpt(context.Background(), WithTestingKnobs(TracerTestingKnobs{Clock: clock}))
+
+	s1 := tr.StartSpan("p", WithRecording(tracingpb.RecordingVerbose))
+	s2 := tr.StartSpan("c", WithParent(s1))
+	extraChildren := 10
+	numChildren := maxRecordedSpansPerTrace + extraChildren
+	payload := &types.Int32Value{Value: int32(1)}
+	for i := 0; i < numChildren; i++ {
+		child := tr.StartSpan(fmt.Sprintf("cc%d", i), WithParent(s2))
+		child.RecordStructured(payload)
+		child.Finish()
+	}
+
+	// We expect recordings from sp and up to the maximum number of spans and
+	// structured records from all spans over the max.
+	rec := s2.FinishAndGetConfiguredRecording()
+	require.Len(t, rec, maxRecordedSpansPerTrace+1)
+	require.Len(t, rec[0].StructuredRecords, extraChildren)
+
+	pl, err := types.MarshalAny(payload)
+	require.NoError(t, err)
+	structuredRecordSize := (&tracingpb.StructuredRecord{Time: now, Payload: pl}).MemorySize()
+	maxNumStructuredRecordings := maxStructuredBytesPerSpan / structuredRecordSize
+	if maxNumStructuredRecordings > numChildren {
+		maxNumStructuredRecordings = numChildren
+	}
+
+	// Since s2's child count exceeded the maximum, we don't expect to see any of
+	// its span recordings in s1. But, we should only see as many of s2's
+	// structured recordings as possible.
+	rec2 := s1.FinishAndGetConfiguredRecording()
+	require.Len(t, rec2, 1)
+	require.Len(t, rec2[0].StructuredRecords, maxNumStructuredRecordings)
+}
+
+// Test that a RecordingStructured parent does not panic when asked to ingest a
+// remote verbose recording. Ingesting a recording of different type is unusual,
+// since children are created with the parent's recording mode, but it can
+// happen if the child's recording mode was changed dynamically.
+func TestRemoteSpanWithDifferentRecordingMode(t *testing.T) {
+	tr := NewTracer()
+	s1 := tr.StartSpan("p", WithRecording(tracingpb.RecordingStructured))
+	s2 := tr.StartSpan("c", WithRemoteParentFromSpanMeta(s1.Meta()), WithRecording(tracingpb.RecordingVerbose))
+	s3 := tr.StartSpan("cc", WithParent(s2), WithRecording(tracingpb.RecordingVerbose))
+	s3.Finish()
+	r := s2.FinishAndGetConfiguredRecording()
+	require.NotPanics(t, func() { s1.ImportRemoteRecording(r) })
+	r2 := s1.FinishAndGetConfiguredRecording()
+	require.Len(t, r2, 1)
+}
+
 type explodyNetTr struct {
 	trace.Trace
 }

